### ISSUE TITLE
[no-extra-boolean-cast] Uncalled Function Checks

### ISSUE THREAD

**Issue by snebjorn** [eslint/eslint#13197]
**What rule do you want to change?**
no-extra-boolean-cast

**Does this change cause the rule to produce more or fewer warnings?**
fewer warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#uncalled-function-checks

```js
  // this check is disallowed, but it's a reasonable function safety check
  if (!!obj.doThing) {
    // safe function call
    obj.doThing()
  }
```

**What does the rule currently do for this code?**
The intended behavior of the rule is to avoid unnecessary casting to a Boolean. But in this scenario I want to indicate that I didn't forget to call the function. I wanted to check if it was defined.

**What will the rule do after it's changed?**
It'll allow the use of `!!` to write something like `if (!!obj.doThing)` to indicate that the coercion is intentional and we're did not forget to invoke the `doThing` function.

The rule should only allow `!!` if it's a function

**Are you willing to submit a pull request to implement this change?**
Yes. but don't have the time at the moment

**Comment by JustFly1984** [comment]
@snebjorn it is better to check functions by type instead of `!!`

```
if (typeof obj.doThing === 'function') {
  obj.doThing()
}
```

**Comment by mdjermanovic** [comment]
Hi @snebjorn, thanks for the issue!

Since we can't reliably know whether `obj.doThing` is supposed to be a function, I think this enhancement is more suitable for [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin).

**Comment by snebjorn** [comment]
That's fair. Gotta get used to where to report linting errors now that TSLint is going away :)
